Streamline turret render traits - part 2

reaperrr opened the pull request:
My idea for making #13619 more flexible is to use integer priority values instead of a single boolean, but even independently from that, our current turret render traits are a bit of a mess, especially when it comes to anything dealing with "aiming" at a target.
This needs to be sorted out first to avoid problems and work-arounds in #13619.

This PR straightens out that mess, so that both #13619 as well as any future additional `WithTurret*Animation` traits should have an easier time to do what they're intended to do without implicit conflicts or compatibility issues.

reaperrr committed:
Replace WithReloadingSpriteTurret with conditions

WithReloadingSpriteTurret was bound to run into conflicts with any WithTurret*Animation traits due to overriding the turret sequence constantly via ITick.
Using (stacked) conditions instead avoids that.

reaperrr committed:
Split *AimAnimation from WithTurretAttackAnimation

These two didn't interact and actually even conflicted when used at the same time, so splitting them is the sensible thing to do.

reaperrr committed:
Make WithTurretAttackAnimation conditional

reaperrr committed:
Make WithTurretAimAnimation conditional

reaperrr committed:
Remove WithSpriteTurret.AimSequence

We can now use WithTurretAimAnimation instead.

pchote commented on OpenRA.Mods.Common/Traits/Render/WithSpriteTurret.cs:
```diff
@@ -119,14 +114,7 @@ protected virtual void DamageStateChanged(Actor self)
-			var sequence = Attack.IsAttacking ? Info.AimSequence : Info.Sequence;
-			DefaultAnimation.ReplaceAnim(sequence);
-		}
+		protected virtual void Tick(Actor self) { }
```
Better to remove `ITick` completely, and if a subclass wants to tick then they can implement the interface themselves.

pchote commented on OpenRA.Mods.Common/Traits/Render/WithTurretAttackAnimation.cs:
```diff
@@ -61,50 +51,45 @@ public WithTurretAttackAnimation(ActorInitializer init, WithTurretAttackAnimatio
 
 		void INotifyAttack.Attacking(Actor self, Target target, Armament a, Barrel barrel)
 		{
-			if (info.DelayRelativeTo == AttackDelayType.Attack)
```
This appears to be almost completely duplicated with `INotifyAttack.PreparingAttack`.
Could you split out a `void NotifyAttack(Actor self, AttackDelayType type)` method that is called from each?

reaperrr replied:
Done.

pchote commented on OpenRA.Mods.Common/Traits/Render/WithTurretAttackAnimation.cs:
```diff
@@ -61,50 +51,45 @@ public WithTurretAttackAnimation(ActorInitializer init, WithTurretAttackAnimatio
-				PlayAttackAnimation(self);
-
-			if (string.IsNullOrEmpty(info.AimSequence) && string.IsNullOrEmpty(info.ReloadPrefix))
+			if (IsTraitDisabled)
```
Not much point having this as an early return.  Just add `!IsTraitDisabled` to the if statement below.

pchote commented on mods/cnc/cursors.yaml:
```diff
@@ -235,3 +235,10 @@ Cursors:
 			Length: 8
 		attackoutsiderange-minimap:
 			Start: 8
+	mouse8.shp: cursor
```
Rebase error?

reaperrr replied:
Oops.

pchote commented on OpenRA.Mods.Cnc/Traits/Render/WithTurretAmmoReloadAnimation.cs:
```diff
@@ -17,34 +17,55 @@
+
+		public override void RulesetLoaded(Ruleset rules, ActorInfo ai)
+		{
+			var matches = ai.TraitInfos<WithTurretAimAnimationInfo>().Count(t => t.Turret == Turret);
```
Style nit: `.Any(t => t.Turret == Turret)` instead of Count != 0.

pchote commented on OpenRA.Mods.Common/Traits/Render/WithTurretAimAnimation.cs:
```diff
@@ -0,0 +1,76 @@
+
+		public override void RulesetLoaded(Ruleset rules, ActorInfo ai)
+		{
+			var turretAttackAnim = ai.TraitInfos<WithTurretAttackAnimationInfo>().Any();
```
Shouldn't this be checking the turret the same way as `WithTurretAmmoReloadAnimationInfo`?

reaperrr commented:
Basically depends on #13538 due to
https://github.com/OpenRA/OpenRA/pull/13538#pullrequestreview-69615367

reaperrr commented:
Updated and rebased.

Note: Added a testcase commit that needs to be removed before merging (makes cnc mobile sam able to attack ground, for easier testing).

reaperrr committed:
De-duplicate some WithTurretAttackAnimation code

reaperrr commented:
Rebased and updated.

Moved testcase here (TD Nod Mobile SAM):
https://github.com/reaperrr/OpenRA/tree/refact-tur-render2-testcase

penev92 commented:
Has merge conflicts again with god-knows-what.

GraionDilach commented:
With #14202.

reaperrr commented:
Rebased.

> With #14202.

And #14425.

reaperrr commented:
Rebased.

abcdefg30 commented:
And needs another rebase.

reaperrr commented:
I guess there's no point in rebasing this until the upgrade rule refactor is merged.

Arular101 commented:
Please don't forget to update the copyright notice year.

reaperrr commented:
Rebased and updated.

reaperrr commented:
Superseded by #14845, #14846 and a future third PR that will contain the rest of the changes from this one.
